countersign.go

```diff
@@ -0,0 +1,299 @@
+func (s *Countersignature) toBeSigned(target any, external []byte) ([]byte, error) {
+	var signProtected cbor.RawMessage
+	signProtected, err := s.Headers.MarshalProtected()
+	if err != nil {
+		return nil, err
+	}
+	return countersignToBeSigned(false, target, signProtected, external)
+}
+
+// countersignToBeSigned constructs Countersign_structure, computes and returns ToBeSigned.
+//
+// Reference: https://datatracker.ietf.org/doc/html/rfc9338#section-3.3
+func countersignToBeSigned(abbreviated bool, target any, signProtected cbor.RawMessage, external []byte) ([]byte, error) {
+	// create a Countersign_structure and populate it with the appropriate fields.
+	//
+	//   Countersign_structure = [
+	//       context : "CounterSignature" / "CounterSignature0" /
+	//                 "CounterSignatureV2" / "CounterSignature0V2" /,
+	//       body_protected : empty_or_serialized_map,
+	//       ? sign_protected : empty_or_serialized_map,
+	//       external_aad : bstr,
+	//       payload : bstr,
+	//       ? other_fields : [+ bstr ]
+	//   ]
+
+	var err error
+	var bodyProtected cbor.RawMessage
+	var otherFields []cbor.RawMessage
+	var payload []byte
+
+	switch t := target.(type) {
+	case SignMessage:
+		if len(t.Signatures) == 0 {
+			return nil, errors.New("SignMessage has no signatures yet")
+		}
+		bodyProtected, err = t.Headers.MarshalProtected()
+		if err != nil {
+			return nil, err
+		}
+		if t.Payload == nil {
+			return nil, ErrMissingPayload
+		}
+		payload = t.Payload
+	case Sign1Message:
+		if len(t.Signature) == 0 {
+			return nil, errors.New("Sign1Message was not signed yet")
+		}
+		bodyProtected, err = t.Headers.MarshalProtected()
+		if err != nil {
+			return nil, err
+		}
+		if t.Payload == nil {
+			return nil, ErrMissingPayload
+		}
+		payload = t.Payload
+		signature, err := encMode.Marshal(t.Signature)
+		if err != nil {
+			return nil, err
+		}
+		signature, err = deterministicBinaryString(signature)
+		if err != nil {
+			return nil, err
+		}
+		otherFields = []cbor.RawMessage{signature}
+	case Signature:
+		bodyProtected, err = t.Headers.MarshalProtected()
+		if err != nil {
+			return nil, err
+		}
+		if len(t.Signature) == 0 {
+			return nil, errors.New("Signature was not signed yet")
+		}
+		payload = t.Signature
+	case Countersignature:
+		bodyProtected, err = t.Headers.MarshalProtected()
+		if err != nil {
+			return nil, err
+		}
+		if len(t.Signature) == 0 {
+			return nil, errors.New("Countersignature was not signed yet")
+		}
+		payload = t.Signature
+	default:
+		return nil, fmt.Errorf("unsupported target %T", target)
+	}
+
+	var context string
+	switch {
+	case len(otherFields) == 0 && abbreviated == false:
+		context = "CounterSignature"
+	case len(otherFields) == 0 && abbreviated == true:
+		context = "CounterSignature0"
+	case len(otherFields) > 0 && abbreviated == false:
+		context = "CounterSignatureV2"
+	case len(otherFields) > 0 && abbreviated == true:
+		context = "CounterSignature0V2"
+	}
```

Comment:
nit: will it be clearer to just use `if` statements
```suggestion
	if len(otherFields) == 0 {
		if abbreviated {
			context = "CounterSignature0"
		} else {
			context = "CounterSignature"
		}
	} else {
		if abbreviated {
			context = "CounterSignature0V2"
		} else {
			context = "CounterSignatureV2"
		}
	}
```